core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -575,9 +584,16 @@ class KafkaApis(val requestChannel: RequestChannel,
       erroneous.foreach{case (tp, data) => partitions.put(tp, data)}
       val unconvertedFetchResponse = fetchContext.updateAndGenerateResponseData(partitions)
 
-      // fetch response callback invoked after any throttling
-      def fetchResponseCallback(bandwidthThrottleTimeMs: Int) {
-        def createResponse(requestThrottleTimeMs: Int): FetchResponse = {
+      // When this callback is triggered, the remote API call has completed.
+      // Record time before any byte-rate throttling.
+      request.apiRemoteCompleteTimeNanos = time.nanoseconds
+
+      def createResponse(throttleTimeMs: Int): FetchResponse = {
+        // If throttling is required, return an empty response.
+        if (throttleTimeMs > 0) {
```

Comment:
Broker's handling of FetchRequest becomes stateful after KIP-227 added support for incremental fetch response. It means that the state in broker will be inconsistent with the state in client if we replace a non-empty FetchResponse with an empty response. More specifically, in `CachedPartitions.updateResponseData()`, state (e.g. highWatermark) will be updated if a partition is assumed to be included in the FetchResponse.

In order to solve this problem, we probably need to change the implementation related to KIP-227. Currently `CachedPartitions.updateResponseData()` will 1) determine the partitions to be included and 2) update the state for those partitions that are included. We probably need to split this into two separate functions. The the state for those partitions should be updated only if the FetchResponse is not throttled.

Another problem is that `clientSensors.quotaSensor.record` has already been called in `recordAndGetThrottleTimeMs()` at this point. It means that we have already assumed that the resource is used to send the FetchResponse at this point. It will cause under-utilization if we only sends an empty FetchResponse after updating the quotaSensor.

The main motivation of this KIP is to address problem caused by ProduceRequest. FetchRequest is small and probably not a concern. If we don't have a simpler way to handle the above two problems, I would recommend not to touch the handleFetchRequest() logic in this patch, i.e. broker still sends the non-empty FetchResponse after the throttle time has passed.

Reply:
Thanks for catching this. I made the following changes to address these points:
1. added getResponseSize() to FetchContext to get the response size (for calculating throttle time) without updating the internal states
2. In case of fetch throttling, unrecord the recorded usage value by recording a negative value of the same quantity. 

I added comments stating more details about these changes.